File: src/main/java/com/fwdekker/randomness/array/ArraySettingsComponent.java
package com.fwdekker.randomness.array;

import com.fwdekker.randomness.SettingsComponent;
import com.fwdekker.randomness.ui.ButtonGroupKt;
import com.fwdekker.randomness.ui.JIntSpinner;
import com.intellij.openapi.ui.ValidationInfo;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.ButtonGroup;
import javax.swing.JCheckBox;
import javax.swing.JPanel;


/**
 * Component for settings of random array generation.
 *
 * @see ArraySettings
 * @see ArraySettingsAction
 */
public final class ArraySettingsComponent extends SettingsComponent<ArraySettings> {
    private JPanel contentPane;
    private JIntSpinner countSpinner;
    private ButtonGroup bracketsGroup;
    private ButtonGroup separatorGroup;
    private JCheckBox spaceAfterSeparatorCheckBox;


    /**
     * Constructs a new {@code StringSettingsComponent} that uses the singleton {@code StringSettings} instance.
     */
    /* default */ ArraySettingsComponent() {
        this(ArraySettings.Companion.getDefault());
    }

    /**
     * Constructs a new {@code StringSettingsComponent} that uses the given {@code StringSettings} instance.
     *
     * @param settings the settings to manipulate with this component
     */
    /* default */ ArraySettingsComponent(final @NotNull ArraySettings settings) {
        super(settings);

        loadSettings();
    }


    @Override
    public JPanel getRootPane() {
        return contentPane;
    }

    /**
     * Initialises custom UI components.
     * <p>
     * This method is called by the scene builder at the start of the constructor.
     */
    private void createUIComponents() {
        countSpinner = new JIntSpinner(1, 1);
    }


    @Override
    public void loadSettings(final @NotNull ArraySettings settings) {
        countSpinner.setValue(settings.getCount());
        ButtonGroupKt.setValue(bracketsGroup, settings.getBrackets());
        ButtonGroupKt.setValue(separatorGroup, settings.getSeparator());
        spaceAfterSeparatorCheckBox.setSelected(settings.isSpaceAfterSeparator());
    }

    @Override
    public void saveSettings(final @NotNull ArraySettings settings) {
        settings.setCount(countSpinner.getValue());

        final String brackets = ButtonGroupKt.getValue(bracketsGroup);
        settings.setBrackets(brackets == null ? ArraySettings.DEFAULT_BRACKETS : brackets);

        final String separator = ButtonGroupKt.getValue(separatorGroup);
        settings.setSeparator(separator == null ? ArraySettings.DEFAULT_SEPARATOR : separator);

        settings.setSpaceAfterSeparator(spaceAfterSeparatorCheckBox.isSelected());
    }

    @Nullable
    @Override
    public ValidationInfo doValidate() {
        return countSpinner.validateValue();
    }
}

File: src/main/kotlin/com/fwdekker/randomness/array/ArraySettingsComponent.kt
package com.fwdekker.randomness.array

import com.fwdekker.randomness.SettingsComponent
import com.fwdekker.randomness.array.ArraySettings.Companion.DEFAULT_BRACKETS
import com.fwdekker.randomness.array.ArraySettings.Companion.DEFAULT_SEPARATOR
import com.fwdekker.randomness.array.ArraySettings.Companion.default
import com.fwdekker.randomness.ui.JIntSpinner
import com.fwdekker.randomness.ui.addChangeListenerTo
import com.fwdekker.randomness.ui.getValue
import com.fwdekker.randomness.ui.setValue
import javax.swing.ButtonGroup
import javax.swing.JCheckBox
import javax.swing.JPanel
import javax.swing.JRadioButton
import javax.swing.event.ChangeEvent


/**
 * Component for settings of random array generation.
 *
 * @param settings the settings to edit in the component
 * @see ArraySettingsAction
 */
@Suppress("LateinitUsage") // Initialized by scene builder
class ArraySettingsComponent(settings: ArraySettings = default) : SettingsComponent<ArraySettings>(settings) {
    override lateinit var rootPane: JPanel private set
    private lateinit var countSpinner: JIntSpinner
    private lateinit var bracketsGroup: ButtonGroup
    private lateinit var separatorGroup: ButtonGroup
    private lateinit var newlineSeparatorButton: JRadioButton
    private lateinit var spaceAfterSeparatorCheckBox: JCheckBox


    init {
        loadSettings()

        newlineSeparatorButton.addChangeListener {
            spaceAfterSeparatorCheckBox.isEnabled = !newlineSeparatorButton.isSelected
        }
        newlineSeparatorButton.changeListeners.forEach { it.stateChanged(ChangeEvent(newlineSeparatorButton)) }
    }


    /**
     * Initialises custom UI components.
     *
     * This method is called by the scene builder at the start of the constructor.
     */
    @Suppress("UnusedPrivateMember") // Used by scene builder
    private fun createUIComponents() {
        countSpinner = JIntSpinner(value = 1, minValue = 1, description = "count")
    }

    override fun loadSettings(settings: ArraySettings) {
        countSpinner.value = settings.count
        bracketsGroup.setValue(settings.brackets)
        separatorGroup.setValue(settings.separator)
        spaceAfterSeparatorCheckBox.isSelected = settings.isSpaceAfterSeparator
    }

    override fun saveSettings(settings: ArraySettings) {
        settings.count = countSpinner.value
        settings.brackets = bracketsGroup.getValue() ?: DEFAULT_BRACKETS
        settings.separator = separatorGroup.getValue() ?: DEFAULT_SEPARATOR
        settings.isSpaceAfterSeparator = spaceAfterSeparatorCheckBox.isSelected
    }

    override fun doValidate() = countSpinner.validateValue()


    override fun addChangeListener(listener: () -> Unit) =
        addChangeListenerTo(
            countSpinner, bracketsGroup, separatorGroup, spaceAfterSeparatorCheckBox,
            listener = listener
        )
}
